### ISSUE TITLE
Run pending WP-Stateless 4.x data migrations across all sites

### ISSUE THREAD

**Issue by curtismchale** [proudcity/wp-proudcity#2889]
## Builds

- [staging](https://console.cloud.google.com/cloud-build/builds;region=global/af140ceb-b972-4e80-8327-10d4847f468b?project=proudcity-1184)
- [prod](https://console.cloud.google.com/cloud-build/builds;region=global/ff0922ed-51c9-4956-a532-6fbafb81769c?project=proudcity-1184)

## Problem

The beta site in the `test` namespace is showing the WP-Stateless upgrade notice:

> Beginning with WP-Stateless 4.0, the method used to store and access plugin data has improved significantly. We've detected that your data still needs to be updated using this new method.

We are running wp-stateless 4.4.1, but the attachment data on our sites is still in the pre-4.0 storage format. This needs to be migrated, and it needs to happen across every site, not one at a time by hand in wp-admin.

## What the migration actually is

Plugin 4.4.1 ships two migrations in `static/migrations/`:

| ID | Description | Batch size | Work per item |
|---|---|---|---|
| `20240219175240` | Update data for Google Cloud files | 1000 | DB only — reads `sm_cloud` postmeta, inserts rows into `wp_sm_files` / `wp_sm_file_sizes` / `wp_sm_file_meta` |
| `20240423174109` | Optimize Compatibility Files | 10 | **Live GCS API call per row** — `$client->get_media()` for each row in `wp_sm_sync` |

Both are **insert-only**. Neither deletes postmeta nor drops a table, and the legacy `sm_cloud` postmeta is left in place after migrating. That makes them non-destructive and re-runnable.

There is WP-CLI support (`lib/cli/class-sm-cli-command.php:296`):

```
wp stateless migrate                # list migrations + status, read-only
wp stateless migrate auto --yes     # run all PENDING migrations
```

`auto` only picks up migrations still marked pending, so it is a no-op on an already-current site.

## Why this can't just go in `entrypoint.sh`

The obvious idea is to drop `wp stateless migrate auto --yes` into `bin/entrypoint.sh` next to the other startup work. That will not work:

1. **Apache isn't running yet.** `Dockerfile:121-122` is `ENTRYPOINT ["/entrypoint.sh"]` + `CMD ["apache2-foreground"]`, and entrypoint ends with `exec "$@"`. The migration does not process batches in-process — `Migrator::start_migration()` queues the batch and calls `dispatch()`, which is a non-blocking loopback POST to `admin-ajax.php` (`lib/ns-vendor/.../wp-background-process.php:108`). Queued before `exec`, there is nothing listening to pick it up. The migration would register as started and then sit there.
2. **entrypoint runs on every pod start**, not just deploys — restarts, crash loops, scale-up. That is a migration attempt per replica against the same database.
3. The cron healthcheck (`sm_batch_process_cron`) is not a fallback — `handle_cron_healthcheck()` also only calls `dispatch()`, so it can't rescue a broken loopback either.

`bin/deploy.sh` is the correct home. It already runs `wp core update-db`, which is exactly the neighbour we want.

## Risks

**Unbounded poll loop.** `_auto_migrate()` (`class-sm-cli-command.php:385`) defaults progress polling to 1s and always blocks, even without `--progress`. `_check_progress()` is a `do { … } while(true)` whose only exit is the migration state option clearing. If the background process never starts (broken loopback) or dies mid-run, it spins forever. In `deploy.sh` that is an indefinitely hung deploy. Any implementation must wrap it in `timeout`.

**Loopback dependency.** Everything above assumes the container can HTTP to its own `admin-ajax.php`. This needs to be confirmed per environment before we trust it. Check:

```bash
wp eval 'echo wp_remote_retrieve_response_code( wp_remote_post( admin_url("admin-ajax.php"), ["timeout"=>10,"body"=>["action"=>"heartbeat"]] ) ) . "\n";'
```

**Deploy duration.** Migration 1 is 1000 attachments per batch with three table inserts each. Migration 2 is one GCS API call per row at batch size 10. On a large media library this is real time and real DB/API load added to every deploy. We need a measurement from a representative site before committing to running it inline.

**Multisite.** `_auto_migrate()` builds its child command without forwarding `--url`, so `auto` mode is broken for multisite installs. Those would need per-migration-ID invocation with an explicit `--url`. Need to confirm whether any of our sites are multisite.

## Do we need to block uploads during migration?

Upstream's notice says "Do not upload, edit, or delete media or files while the optimization process is underway." Reading the code, a hard freeze looks like more than we need:

- Nothing in the plugin gates uploads. The only concurrency handling is an admin nag (`class-bootstrap.php:194`).
- Both migrations are insert-only and leave the old postmeta intact, so a concurrent upload cannot destroy data.
- The new tables carry `UNIQUE KEY post_id_name` / `post_id_size` / `post_id_key` (`class-db.php:177-203`) and each attachment is processed inside a transaction with rollback. A race produces a rejected insert, not a duplicate or a half-written row.
- Migration 1 snapshots the max attachment ID at init and walks **descending** (`ID < offset`), so anything uploaded during the run is above the starting offset and is never visited. That is correct — under 4.x runtime, new uploads already write new-format rows directly.

What still argues for a quiet window rather than mid-day:

- Migration 2's per-row GCS call will contend with normal upload traffic for API quota and is the slow part.
- **Edits and deletes of existing media mid-run are the untested path.** A delete between batch selection and `process_item()` leaves an orphan row; an edit can leave the legacy postmeta and the new row disagreeing. Neither is data loss, but neither is verified.

Proposed position: no upload lock, but run during a low-traffic window and verify on beta first.

## Proposed plan

**Phase 1 — measure and verify on beta (test namespace)**

- [ ] Confirm the loopback check above returns 200 inside a real pod
- [ ] Take a DB backup
- [ ] Record attachment count and `wp_sm_sync` row count
- [ ] Run `wp stateless migrate` to capture the pending list and status
- [ ] Run `wp stateless migrate auto --yes --progress=10` by hand, timing the whole run
- [ ] Confirm the notice clears and media still resolves (front end + wp-admin media library)
- [ ] Spot-check `wp_sm_files` row counts against attachment count

**Phase 2 — decide the rollout vehicle based on Phase 1 timing**

If the run is fast (single-digit minutes) on a representative site, add it to `bin/deploy.sh` after `wp core update-db`:

```sh
# Run any pending WP-Stateless data migrations (the 4.x storage format change).
# Batches are handled by a background loopback request to admin-ajax.php, so this has
# to run while Apache is serving -- that's why it lives here and not in entrypoint.sh.
# `auto` only picks up migrations still marked pending, so this is a no-op once a site
# is current. The timeout is a guard: `migrate auto` polls the migration state option
# in an unbounded loop, so a stalled background process would otherwise hang the deploy.
if wp plugin is-active wp-stateless; then
    echo "Running pending WP-Stateless migrations"
    timeout 1800 wp stateless migrate auto --yes || \
        echo "WP-Stateless migration did not finish cleanly, check 'wp stateless migrate'" >&2
fi
```

If the run is slow, do **not** put it inline in `deploy.sh` — `deploy.sh` runs synchronously and would stall the release. Instead do a one-off pass per namespace, scheduled in a low-traffic window:

```bash
kubectl exec -n <ns> <pod> -- wp stateless migrate auto --yes --progress=10
```

**Phase 3 — roll out**

- [ ] Backup before each site (or confirm the existing backup cadence covers it)
- [ ] Run against remaining sites, low-traffic window
- [ ] Verify the notice cleared and media resolves on each

## Open questions for approval

1. Inline in `deploy.sh`, or a separate scheduled one-off pass? Depends on Phase 1 timing.
2. Do we want an upload freeze, or is "low-traffic window, no lock" acceptable given the analysis above?
3. What is the acceptable `timeout` ceiling for a deploy-blocking migration?
4. Are any of our sites multisite (affects `auto` mode)?

Developer notes: `~/Documents/developers/Github Issue Notes/`

**Comment by curtismchale** [comment]
Ran and verified all sites.

Still need to figure out a method to make sure we don't miss migrations during plugin upgrades.
